src/main/java/org/truffleruby/core/array/ArrayBuilderNode.java

```diff
@@ -26,520 +31,236 @@
-                return store;
-            }
-
-            CompilerDirectives.transferToInterpreterAndInvalidate();
-            return replace(new ObjectArrayBuilderNode(expectedLength)).
-                    appendArray(ArrayUtils.box((int[]) store), index, array);
-        }
-
-        @Override
-        public Object appendValue(Object store, int index, Object value) {
-            if (store instanceof int[] && value instanceof Integer) {
-                ((int[]) store)[index] = (int) value;
-                return store;
-            } else {
+            if (lengthProfile.profile(length > expectedLength)) {
                 CompilerDirectives.transferToInterpreterAndInvalidate();
-                return appendValueFallback(store, index, value, expectedLength);
+                replaceNodes(strategy, length);
             }
+            return strategy.newArray(length).getArray();
         }
 
-        @Override
-        public Object finish(Object store, int length) {
-            return store;
+        public boolean replacement(ArrayStrategy strategy, int newLength) {
+            final StartNode newNode = new StartNode(strategy, Math.max(expectedLength, newLength));
+            this.replace(newNode);
+            return (strategy != this.strategy);
         }
-
     }
 
-    private static class LongArrayBuilderNode extends ArrayBuilderNode {
-
-        private final int expectedLength;
-        private final ConditionProfile otherLongStoreProfile = ConditionProfile.createBinaryProfile();
+    @ImportStatic(ArrayGuards.class)
+    public abstract static class AppendOneNode extends ArrayBuilderBaseNode {
 
-        public LongArrayBuilderNode(int expectedLength) {
-            this.expectedLength = expectedLength;
+        public static AppendOneNode create() {
+            return AppendOneNodeGen.create();
         }
 
-        @Override
-        public Object start() {
-            return new long[expectedLength];
-        }
+        public abstract Object executeAppend(Object array, int index, Object value);
 
-        @Override
-        public Object start(int length) {
-            if (length > expectedLength) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                return restart(length, length + " > " + expectedLength + " (expected)");
+        @Specialization(guards = { "strategy.matchesStore(array)", "strategy.accepts(value)" },
+                        limit = "1")
+        public Object appendCompatibleType(Object array, int index, Object value,
+                @Cached("ofStore(array)") ArrayStrategy strategy,
+                @Cached("createBinaryProfile()") ConditionProfile expandProfile) {
+            ArrayMirror mirror = strategy.newMirrorFromStore(array);
+            if (expandProfile.profile(index >= mirror.getLength())) {
+                final int capacity = Math.max(mirror.getLength(), index + 1);
```

Comment:
Why not `ArrayUtils.capacityForOneMore()` here?

Reply:
That requires a `RubyContext`, and I had some problems with `RubyBaseNode`'s specification of the ruby type system causing deopts. I guess we could pass in the context from the proxy node easily enough.

Reply:
ArrayBuilderBaseNode should probably just be a RubyBaseNode and then this should work out of the box.

> I had some problems with RubyBaseNode's specification of the ruby type system causing deopts

That sounds very weird and worth fixing.

> I guess we could pass in the context from the proxy node easily enough.

Yeah, that should work too.

Reply:
So I've looked at the issue with DSL code generation a bit more. If an argument to the abstract execute method is an `int` then we the DSL doesn't produce sensible code. Everything works if I change the `int` to an `Object`, or if I use a node type that doesn't have `@TypeSystemReference(RubyTypes.class)`.

Reply:
What's "not sensible code"? Can you report a bug to Truffle then?